lib/spack/spack/spec.py

```diff
@@ -1826,16 +1846,19 @@ def to_json(self, stream=None, hash=ht.dag_hash):
 
     @staticmethod
     def from_node_dict(node):
-        name = next(iter(node))
-        node = node[name]
-
         spec = Spec()
+        if 'name' in node.keys():
+            # New format
+            name = node['name']
+        else:
+            # Old format
+            name = next(iter(node))
+            node = node[name]
+        for h in ht.hashes:
+            setattr(spec, h.attr, node.get(h.name, None))
```

Comment:
Definitely not asking for a refactor here, but in general I wonder if when we upgrade formats we should have separate functions to read different formats.